File: example/cpp/src/hardware/arm/return_zero.cpp
#include <cmath>
#include <iostream>
#include <thread>

#include "airbot_hardware/executors/executor.hpp"
#include "airbot_hardware/handlers/arm.hpp"

using namespace std::chrono_literals;
using namespace airbot::hardware;

bool all_zero(const std::array<double, 7>& pos) {
  for (const auto& p : pos)
    if (std::abs(p) > 1e-3) return false;
  return true;
}

int main() {
  auto executor = airbot::hardware::AsioExecutor::create(8);
  auto arm = airbot::hardware::Arm<7>::create<MotorType::OD, MotorType::OD, MotorType::OD, MotorType::DM, MotorType::DM,
                                              MotorType::DM, EEFType::G2, MotorType::DM>();
  if (!arm->init(executor->get_io_context(), "can0", 250)) {
    std::cerr << "Failed to initialize arm" << std::endl;
    return 1;
  }
  arm->enable();
  arm->set_param("arm.control_mode", static_cast<uint32_t>(MotorControlMode::PVT));
  arm->set_param("eef.control_mode", static_cast<uint32_t>(MotorControlMode::PVT));

  auto now = std::chrono::steady_clock::now();
  while (!all_zero(arm->state().pos) && std::chrono::steady_clock::now() - now <= 10s) {
    arm->pvt({0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0},
             {M_PI / 10, M_PI / 10, M_PI / 10, M_PI / 10, M_PI / 10, M_PI / 10, M_PI / 10});
    std::this_thread::sleep_for(1ms);
  }

  arm->disable();
  arm->uninit();
  return 0;
}

File: example/python/hardware/arm/return_zero.py
import airbot_hardware_py as ah
import time
import math


def all_zero(pos):
    """Check if all positions are close to zero (within 1e-3)"""
    return all(abs(p) < 1e-3 for p in pos)


def main():
    executor = ah.create_asio_executor(8)
    io_context = executor.get_io_context()

    arm = ah.Play.create(
        ah.MotorType.OD,
        ah.MotorType.OD,
        ah.MotorType.OD,
        ah.MotorType.DM,
        ah.MotorType.DM,
        ah.MotorType.DM,
        ah.EEFType.G2,
        ah.MotorType.DM,
    )

    if not arm.init(io_context, "can0", 250):
        raise RuntimeError("Failed to initialize arm")

    arm.enable()
    arm.set_param("arm.control_mode", ah.MotorControlMode.PVT)

    now = time.time()
    while not all_zero(arm.state().pos) and time.time() - now <= 10:
        arm.pvt(
            [0.0, 0.0, 0.0, 0.0, 0.0, 0.0],  # target positions
            [
                math.pi / 10,
                math.pi / 10,
                math.pi / 10,
                math.pi / 10,
                math.pi / 10,
                math.pi / 10,
            ],  # velocities
        )
        time.sleep(0.001)  # 1ms sleep

    arm.disable()
    arm.uninit()


if __name__ == "__main__":
    main()
